Revise the doc for Streaming Response (#3847)
line/armeria: site/src/pages/docs/advanced-streaming-backpressure.mdx

# Sending a streaming response with back pressure

Suppose we want to serve a static file whose size is larger than your available memory. If we load the file into the memory at once, we will definitely get an `OutOfMemoryError`.

To prevent such error and send large files without consuming too much memory:

1. Divide the file into chunks.
2. Load the first chunk into memory.
3. Send the chunk to the client.
4. Wait until the chunk is written to the sending socket buffer. (See below for explanation.)
5. Load the second chunk and repeat the steps 3 and 4 until we send all chunks.

Waiting for the chunk to be written is to avoid loading into memory when the client is not ready to receive. This is called __back pressure__. 

> See [Let’s Play with Reactive Streams on Armeria - 1](https://engineering.linecorp.com/en/blog/reactive-streams-armeria-1/) to learn back pressure and what happens when an `OutOfMemoryError` is raised.

Reviewer: How about 
```suggestion
See [Let’s Play with Reactive Streams on Armeria - 1](https://engineering.linecorp.com/en/blog/reactive-streams-armeria-1/) to learn back pressure and when an `OutOfMemoryError` is raised.
```
Because I wanted to say that an `OutOfMemoryError` is raised if we don't use back pressure.
We all know what happens when an `OutOfMemoryError` is raised. 😄

Author: @minwoox , 
Hope this fix is as you expected
https://github.com/line/armeria/pull/3847/commits/1ad1e22b7fd50f4cece2d1a566ce2c53d35941bb

      .service("/big_file.dat", bigFile.asService())
      .build();
```
The <type://HttpFile> loads chunks one by one and send them to the client with backpressure.

> ## Sending a streaming response using <type://HttpResponseWriter>

Reviewer: So it's not a good idea to use a link in the subtitle?

Author: Some sites have clickable headings which points to themselves so users can copy a link straight away from the URL field of a browser. Considering that many tools these days add links to headings  automatically (like  Armeria site providing a permalink) by SSG, adding links to a heading doesn't seems like a good practice. Although Armeria overcomes this and generates an anchor that works perfectly well, having the term "type" attached to the HttpResponseWriter may be confusing in terms of accsesibility.

Reviewer: That totally makes sense. We have a link for the type in the section anyway. 😄

To send large data other than files such as database, you need to implement back pressure yourself. Let's start off with implementing a minimal <type://HttpFile>.

> Prepare to send a streaming response with <type://HttpResponseWriter> and <type://HttpResponse#streaming()>.

Reviewer: I wanted to say that `HttpResponseWriter` is returned when you call `HttpResponse.streaming()` which is:
`HttpResponseWriter writer = HttpResponse.streaming();`
So how about:
```suggestion
Prepare to send a streaming response with <type://HttpResponseWriter> returned from <type://HttpResponse#streaming()>.
```

Reviewer 2: returned by?

Reviewer: Oh, that's correct. 😄

...
private HttpData produceChunk(int index) {
    // Divide the file by the pre-defined chunk size(e.g. 8192 bytes)

>     // and read it using index. 

Reviewer: revert?

    // If index is 0, 0 to 8192 bytes from the file is read.
    // If index is 1, 8193 to 16384 is read and so on.
}
```

> With the code above, the server would still encounter `OutOf MemoryError`. We still need to take care of preventing loading data chunks into memory before a chunk is sent to the client. To solve the problem, implement back pressure with <type://StreamWriter#whenConsumed()>:

Reviewer: ```suggestion
With the code above, the server would encounter `OutOfMemoryError`. We still need to take care of preventing loading data chunks into memory before a chunk is sent to the client. To solve the problem, implement back pressure with <type://StreamWriter#whenConsumed()>:
```
Perhaps, it's better to use `still` only once?

});
```

> <type://StreamWriter#whenConsumed()> returns a `CompletableFuture` that is complete when the chunk written to the <type://HttpResponseWriter> is finally written to the socket. This enables you to add the next task by adding a callback (`thenRun()` in the code example). The next task in the example is set to producing the next chunk.

Reviewer: ```suggestion
<type://StreamWriter#whenConsumed()> returns a `CompletableFuture` that is complete when the chunk written to the <type://HttpResponseWriter> is finally written to the socket. This enables you to add the next task by adding a callback (`thenRun()` in the code example). The next task in the example is set to produce the next chunk.
```

    });
}
```

> So far, we have implemented a simple version of <type://HttpFile>. Now, we can implement a streaming response with back pressure for any type of source (e.g. database) by simply changing the `produceChunk()` method to fetch data from the  source.

Reviewer: ```suggestion
So far, we have implemented a simple version of <type://HttpFile>. Now, we can implement a streaming response with back pressure for any type of source (e.g. database) by simply changing the `produceChunk()` method to fetch data from the source.
```
